docs/05-internals/design/principles/feature-toggles.md

```diff
@@ -0,0 +1,281 @@
+<!---
+Copyright © 2015-2019 the contributors (see Contributors.md).
+
+This file is part of Knora.
+
+Knora is free software: you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as published
+by the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+Knora is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public
+License along with Knora.  If not, see <http://www.gnu.org/licenses/>.
+-->
+
+# Feature Toggles
+
+For an overview of feature toggles, see
+[Feature Toggles (aka Feature Flags)](https://martinfowler.com/articles/feature-toggles.html).
+The design presented here is partly inspired by that article.
+
+## Requirements
+
+- It should be possible to turn features on and off by:
+
+  - changing a setting in `application.conf`
+  
+  - sending a particular HTTP header value with an API request
+
+  - (in the future) using a web-based user interface to configure a
+    feature toggle service that multiple subsystems can access
+    
+    
+- Feature implementations should be produced by factory classes,
+  so that the code using a feature does not need to know
+  about the toggling decision.
+  
+- Feature factories should use toggle configuration taken
+  from different sources, without knowing where the configuration
+  came from.
+  
+- An HTTP response should indicate which features are turned
+  on.
```

Comment:
Currently it only returns enabled toggles. I was thinking that if a toggle is turned off by default because the feature isn't ready to use yet, there wouldn't be much point in returning it. But maybe it would be clearer and simpler to return everything. What do you think?

Reply:
I think returning everything might not be a good idea. Since the lifetime of toggles is a year, we might have many (>10 even) toggles in place that are turned off but still available. IMO, Returning them would confuse the user.